### DISCUSSION TITLE
main.py doesn't auto start

### DISCUSSION THREAD

**Discussion by Johnl52** [micropython/micropython#10591] [RP2040 / Pico]
I have a pico with a OLED display, a BME280 temp sensor and a CMPS12 compass module. The program name is main.py and I'm using Thonny. I have the temp, humidity, pressure and compass heading displaying on the display. I first had just the temp, humidity and pressure when I unplug the USB cable then plug it back in the code would run. After I added the code for the compass it does not auto start. I have to reconnect to the pico and press F5. It wouldn't start by pressing the reset on the pico either. I loaded the first program I had back in and that works. What is wrong with this program?

Thanks

```
from machine import Pin, I2C
from ssd1306 import SSD1306_I2C
from time import sleep
import bme280

i2c=I2C(0,sda=Pin(0), scl=Pin(1), freq=400000)
oled = SSD1306_I2C(128, 64, i2c)
bme = (bme280.BME280(i2c=i2c))

def read_bytes(i2cAddr, regAddr, buf):
    cmd    = bytearray(1)
    cmd[0] = regAddr & 0xff
    i2c.writeto(i2cAddr, cmd, False)
    i2c.readfrom_into(i2cAddr, buf)
    

def get_heading(hires):
    """ Returns heading with or w/o tilt compensation and/or calibration,
        if available.
        NOTE: The CMPS12 has built-in tilt compensation and is pre-calibra-
        ted, therefore the parameters "tilt" and "calib" are only for
        compatibility reasons and have no effect. With "hires"=True, the
        precision is 3599/360°, otherwise 255/360°.
    """
    ADDRESS_CMPS12 = const(0x60)
    REG_BEARING_8BIT = const(0x01)
    REG_BEARING_16BIT_HB = const(0x2)
    
    hd = 1
    if hires:
      buf = bytearray(2)
      read_bytes(ADDRESS_CMPS12, REG_BEARING_16BIT_HB, buf)
      #print("buf=",buf)
      hd  = ((buf[0] << 8) | buf[1]) /10
    else:
      buf = bytearray(1)
      read_bytes(ADDRESS_CMPS12, REG_BEARING_8BIT, buf)
      hd  = buf[0]/255 *360
    return hd

while True:
    temp, pressure, humidity = bme.read_compensated_data()
    temp_f = ((temp * 1.8) / 100) + 32
    press_Hg = (pressure/256) / 3386.39
    humd = humidity/1024
    reading = 'Temperature: ' + f'{(temp_f): .2f}' + '   Humidity: ' + f'{(humd): .2f}' + '   Pressure: ' + f'{(press_Hg): .2f}'
    print(reading)
    head = get_heading(True)
    print ("heading=",head)
    oled.fill(0)
    # Show values (data, x, y)
    oled.text("Temp: ", 1, 5 )
    oled.text(f'{(temp_f): .2f}', 50, 5)
    oled.text("Press: ", 1, 20 )
    oled.text(f'{(press_Hg): .2f}', 50, 20)
    oled.text("Humid: ", 1, 35 )
    oled.text(f'{(humd): .2f}', 50, 35)
    oled.text("Head: ", 1, 50 )
    oled.text(f'{(head): .2f}', 50, 50)
    oled.show()
    sleep(1)
```

**Comment by cve2022** [comment]
Delete main.py, save this script with a different name and import it on the REPL terminal. It can print error messages that could explain why it is not working.

You can later create a main.py with the import script_name on it.

**Reply by jimmo** [reply]
Good tip from @cve2022

I'm not sure what's going wrong though that Thonny isn't showing you error messages from the device, that seems important. You should be able to hit run and see any debug output on the console.

**Comment by Johnl52** [comment]
@cve2022 I copied the code to the REPL, wasn't sure if I imported it right as you mentioned, I copied the main.py code, renamed the file, then pasted the code into the REPL, it worked fine no errors. It was also printing out the block of comments in the def get_heading(hires): function but I put a # in front of each line and that stopped it from printing in the REPL. It does display a bold blue "1", I'm not sure where that's coming from but it doesn't seem to be an issue. I then copied that back to a main.py file, ran it then unplugged it, plugged it back in and it doesn't start until I press F5. I would think if there was a code problem I would get an error when pressing F5 as I have made some code errors and it did. I fairly new at using micropython.

Thanks both for your comments

**Comment by Johnl52** [comment]
I have determined it has something to do with the "get_heading(hires)" function. I commented out line 48 "head = get_heading(True)" and had to comment out a couple of print lines that use "head" but then it works, both with removing and plugging the USB cable back in and with a battery I have connected to VSYS. Leaving the print statements commented if I just uncomment the call to get_heading, line 48, then it doesn't work.

**Reply by robert-hh** [reply]
You must execute  statements like `ADDRESS_CMPS12 = const(0x60)` only once in your code, because is replaces `ADDRESS_CMPS12` by a literal number on the first call. On the second call the code tries to assign to a constant, which fails. Usually, statements like `ADDRESS_CMPS12 = const(0x60)` are placed at the start of the file to ensure, that they are executed only once during import.

**Comment by Johnl52** [comment]
Good point @robert-hh, I'll remember that for future projects. But I'm still having the autorun issue when I call get_heading.

Thanks 
John

**Reply by robert-hh** [reply]
You could use a try/except clause and write at least the exception reason to a file.
